open_spiel/games/kriegspiel.h

```diff
@@ -0,0 +1,276 @@
+// Copyright 2019 DeepMind Technologies Ltd. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef OPEN_SPIEL_GAMES_KRIEGSPIEL_H_
+#define OPEN_SPIEL_GAMES_KRIEGSPIEL_H_
+
+#include <array>
+#include <cstring>
+#include <map>
+#include <memory>
+#include <string>
+#include <vector>
+#include <open_spiel/fog/observation_history.h>
+
+#include "open_spiel/abseil-cpp/absl/container/flat_hash_map.h"
+#include "open_spiel/abseil-cpp/absl/algorithm/container.h"
+#include "open_spiel/games/chess/chess_board.h"
+#include "open_spiel/games/chess.h"
+#include "open_spiel/spiel.h"
+#include "open_spiel/spiel_utils.h"
+
+// Kriegpsiel - imperfect information version of chess:
+// https://en.wikipedia.org/wiki/Kriegspiel
+//
+// Parameters:
+//   "board_size"         int     Number of squares in each row and column
+//                                (default: 8)
+//   "fen"                string  String describing the chess board position in
+//                                Forsyth-Edwards Notation. The FEN has to match
+//                                the board size. Default values are available
+//                                for board sizes 4 and 8.
+//   "3_fold_repetition"  bool    Whether 3-fold repetition rule should be
+//                                automatically enforced (default: false)
+//   "50_move_rule"       bool    Whether 50 move rule should be automatically
+//                                enforced (default: false)
+
+namespace open_spiel {
+namespace kriegspiel {
+
+// Constants.
+inline constexpr int NumPlayers() { return 2; }
+inline constexpr double LossUtility() { return -1; }
+inline constexpr double DrawUtility() { return 0; }
+inline constexpr double WinUtility() { return 1; }
+
+// See action encoding below.
+inline constexpr int NumDistinctActions() { return 4672; }
+
+// https://math.stackexchange.com/questions/194008/how-many-turns-can-a-chess-game-take-at-maximum
+inline constexpr int MaxGameLength() { return 17695; }
```

Comment:
I think this is higher

Reply:
@michalsustr You're right. It is implemented in a way that attempting an illegal move is still considered a move. I can put a loose upper bound there. However, with the threefold repetition and 50 move rule turned off, the gamecan hypothetically go on forever.